File: ruby_edition/cm_challenge/absences.rb
require_relative './api'
require_relative './members'
require_relative './absences_helper'

module CmChallenge
  class Absences
    extend AbsencesHelper
    class << self
      def to_ical
        generate_ical_data_from_absences(all_with_names)
      end

      def generate_ical_file(file_name: 'absences.ics', data: to_ical)
        write_data_to_file(data: data, file_name: file_name)
      end

      def all_with_names
        absences = Api.absences
        absences.each do |absence|
          absentee = get_absentee_from_members(absentee_id: absence[:user_id])
          absence[:employee_name] = absentee[:name]
        end

        absences
      end

      def find_employee_absences(employee_id:)
        all_with_names.select { |absence| absence[:user_id] == employee_id }
      end

      def employee_vacation_status(employee:, status:)
        return "#{employee[:name]} is on #{status} leave" if is_absent?(employee_id: employee[:user_id], absence_type: status)

        "#{employee[:name]} is currently NOT absent"
      end

      def between(start_date:, end_date:)
        all_with_names.select do |absence| format_date(date: absence[:start_date]) >= format_date(date: start_date) &&
            format_date(date: absence[:end_date]) <= format_date(date: end_date)
        end
      end

      private

      def get_absentee_from_members(absentee_id:)
        Members.find(user_id: absentee_id)
      end

      def is_absent?(employee_id:, absence_type:)
        employee_absences = find_employee_absences(employee_id: employee_id)
        employee_absences.any? {|absence| format_date(date: absence[:end_date]) >= Date.today && absence[:type] == absence_type}
      end
    end
  end
end

File: js_edition/src/absences.js
import moment from 'moment';
import ical from 'ical-generator';
import { writeFileSync } from 'fs';

import { absences, members } from './api';

export const allAbsencesWithNames = async () => {
  try {
    let allAbsences = await absences();
    let transformedAbsences = [];

    for (let absence of allAbsences){
      let member = await getMemberById(absence.userId);
      absence.employeeName = member.name;
      transformedAbsences.push(absence);
    }

    return transformedAbsences;
  } catch (error) {
    throw new Error(error);
  }
};

export const toIcal = async () => {
  try {
    const all_absences = await allAbsencesWithNames();
    const events = [];

    for (let absence of all_absences) {
      const event = {
        start: moment(absence.startDate),
        end: moment(absence.endDate),
        timestamp: moment(),
        description: `${absence.employeeName} is absent due to ${absence.type}.`,
        summary: `${absence.employeeName}'s absence details`,
        organizer: 'Manager <mail@example.com>'
      };

      events.push(event);
    }
    return ical({
      domain: 'mydomain.net',
      prodId: '//mydomain.com//ical-generator//EN',
      events
    }).toString();
  } catch (error) {
    throw new Error(error);
  }
};

export const writeIcalDataToFile = async (filename = 'absences', data = toIcal) => {
  try {
    const absences = await  data();
    writeFileSync(`${__dirname}/${filename}.ics`, absences);
  } catch (error) {
    throw new Error(error)
  }

};

export const employeeAbsenceStatus = async (employee, absences = allAbsencesWithNames) => {
  const status = {
    vacation: (member) => `${member} is on vacation`,
    sickness: (member) => `${member} is sick`,
    default: (member) => `${member} is currently NOT absent`,
  };

  try {
    const allAbsences = await absences();
    const absence = employeeAbsence(employee.userId, allAbsences);

    // Return absence type if employee has any absence or null otherwise
    const type = absence && absence.type || null;

    return status[type || 'default'](employee.name)
  } catch (error) {
    throw new Error(error);
  }
};

const getMemberById = async (id) => {
  try {
    let allMembers = await members();
    return allMembers.find((member) => member.userId === id);
  } catch (error) {
    throw new Error(error);
  }
};

const employeeAbsences = (employeeId, absences) => {

  return absences.filter((absence) => absence.userId === employeeId);
};

const employeeAbsence = (employeeId, absences) => {
  try {
    const allEmployeeAbsences = employeeAbsences(employeeId, absences);

    return allEmployeeAbsences.find((absence) => moment(absence.endDate) >= new Date());
  } catch (error) {
    throw new Error(error);
  }
};
